### PR TITLE
feat: enable compatibility of `xstate-vue` with Vue 3

### LINKED ISSUE #1252: @xstate/vue not working with vue beta
**Description**
@xstate/vue is not working with vue beta. The module still use @vue/composition-api as a dependency. Should we drop it as vue has enter beta ?

**Actual Result**
when I use the example in the readme, I get an error like 
```
Uncaught SyntaxError: The requested module '/@modules/@xstate/vue/lib/index.js' does not provide an export named 'useMachine'
```
**Reproduction**
Create a new project with `vue-cli` or `vite`
To use vue 3 (with composition api) with `vue-cli` you need to do : `vue add vue-next`
Then just write the example in the readme file, you should get the error.

I am pretty new to xstate and all the concepts behind it, so I am not sure how to start.

Any ideas ? Thanks a lot !

### PR DESCRIPTION
**Pull request by sarahdayan**
This PR enables compatibility of `xstate-vue` with Vue 3. It removes the peer dependency with `@vue/composition-api`. **This creates a breaking change**.

I had trouble running tests on my machine, I'll see how things look like in the CI and will take action to rectify if necessary.

cc @Andarist if you want to take a look 🙂

closes #1252

### CODE REVIEW

**Comment by davidkpiano** [comment]
Thanks, Sarah! I think the way the tests were previously set up aren't playing well with the Vue 3 typings:

```
Summary of all failing tests
 FAIL  packages/xstate-vue/test/useMachine.test.ts
  ● Test suite failed to run

    packages/xstate-vue/test/useMachine.test.ts:49:47 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.
        Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' provides no match for the signature 'new (...args: any[]): Vue'.

    49     const { getByText, getByTestId } = render(UseMachine);
                                                     ~~~~~~~~~~
    packages/xstate-vue/test/useMachine.test.ts:59:47 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.

    59     const { getByText, getByTestId } = render(UseMachine, {
                                                     ~~~~~~~~~~
    packages/xstate-vue/test/useMachine.test.ts:71:47 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.

    71     const { getByText, getByTestId } = render(UseMachine, {
                                                     ~~~~~~~~~~
    packages/xstate-vue/test/useMachine.test.ts:81:27 - error TS2769: No overload matches this call.
      Overload 1 of 3, '(component: VueClass<Vue>, options?: ThisTypedMountOptions<Vue> | undefined): Wrapper<Vue>', gave the following error.
        Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
          Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.
      Overload 2 of 3, '(component: ComponentOptions<Vue, DefaultData<Vue>, DefaultMethods<Vue>, DefaultComputed, PropsDefinition<Record<string, any>>, Record<...>>, options?: ThisTypedMountOptions<...> | undefined): Wrapper<...>', gave the following error.
        Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'ComponentOptions<Vue, DefaultData<Vue>, DefaultMethods<Vue>, DefaultComputed, PropsDefinition<Record<string, any>>, Record<...>>'.
          Types of property 'computed' are incompatible.
            Type '{ <T>(getter: ComputedGetter<T>): ComputedRef<T>; <T>(options: WritableComputedOptions<T>): WritableComputedRef<T>; }' is not assignable to type 'Accessors<DefaultComputed>'.
              Index signature is missing in type '{ <T>(getter: ComputedGetter<T>): ComputedRef<T>; <T>(options: WritableComputedOptions<T>): WritableComputedRef<T>; }'.
      Overload 3 of 3, '(component: FunctionalComponentOptions<Record<string, any>, PropsDefinition<Record<string, any>>>, options?: MountOptions<Vue> | undefined): Wrapper<...>', gave the following error.
        Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'FunctionalComponentOptions<Record<string, any>, PropsDefinition<Record<string, any>>>'.
          Property 'functional' is missing in type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' but required in type 'FunctionalComponentOptions<Record<string, any>, PropsDefinition<Record<string, any>>>'.

    81     const wrapper = mount(UseMachine, { localVue });
                                 ~~~~~~~~~~

      node_modules/vue/types/options.d.ts:132:3
        132   functional: boolean;
              ~~~~~~~~~~
        'functional' is declared here.
    packages/xstate-vue/test/useMachine.test.ts:90:14 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.

    90       render(UseMachineNoExtraOptions);
                    ~~~~~~~~~~~~~~~~~~~~~~~~

 FAIL  packages/xstate-vue/test/useService.test.ts
  ● Test suite failed to run

    packages/xstate-vue/test/useService.test.ts:36:39 - error TS2345: Argument of type '{ components: { UseService: typeof UseService; }; template: string; props: string[]; }' is not assignable to parameter of type 'VueClass<Vue>'.
      Type '{ components: { UseService: typeof UseService; }; template: string; props: string[]; }' is not assignable to type 'new (...args: any[]) => Vue'.
        Type '{ components: { UseService: typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue"); }; template: string; props: string[]; }' provides no match for the signature 'new (...args: any[]): Vue'.

    36     const { getAllByTestId } = render(twoServices, {
                                             ~~~~~~~~~~~
    packages/xstate-vue/test/useService.test.ts:57:49 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.
        Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' provides no match for the signature 'new (...args: any[]): Vue'.

    57     const { getByTestId, updateProps } = render(UseService, {
                                                       ~~~~~~~~~~

 FAIL  packages/xstate-vue/test/useFSM.test.ts
  ● Test suite failed to run

    packages/xstate-vue/test/useFSM.test.ts:14:47 - error TS2345: Argument of type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to parameter of type 'VueClass<Vue>'.
      Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' is not assignable to type 'new (...args: any[]) => Vue'.
        Type 'typeof import("/Users/davidkpiano/Code/xstate-v4/packages/xstate-vue/node_modules/vue/dist/vue")' provides no match for the signature 'new (...args: any[]): Vue'.

    14     const { getByText, getByTestId } = render(UseFSM);
                                                     ~~~~~~
```

I can help look into these. Might just need some refactoring (instead of inlining the test components).

**Comment by pearofducks** [comment]
Just a heads up - [this component](https://github.com/davidkpiano/xstate/blob/master/packages/xstate-vue/src/xstate-machine.ts) isn't Vue 3 compatible at the moment.

File: .changeset/tame-rockets-watch.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@xstate/vue': minor
+---
+
+Enabled compatibility with Vue 3 and removed compatibility with @vue/composition-api.
```
Comment (Andarist) [inline .changeset/tame-rockets-watch.md:5]:
```suggestion
Added support for Vue 3 that has builtin support for the composition API. This means that this package will no longer work with [`@vue/composition-api`](https://github.com/vuejs/composition-api) package. If you are still using Vue 2 you should continue using `@xstate/vue@^0.3.0`.
```

**Comment by Andarist** [comment]
@sarahdayan could you merge master to your branch? It seems that CI was not running for external contributions - I've fixed that now.

**Comment by guilipan** [comment]
any progress here?

**Comment by davidkpiano** [comment]
> any progress here?

Patience 🙏

**Comment by TheDutchCoder** [comment]
Any movement on this?
It's kind of a critical thing for us Vue devs :)

File: package.json [hunk 2]
```diff
@@ -58,6 +58,7 @@
     "lint-staged": "^8.2.1",
     "markdown-it-codesandbox-embed": "^0.1.0",
     "mermaid": "^8.2.3",
+    "patch-package": "^6.2.2",
```
Comment (Andarist) [inline package.json:61]: I have not used this in the end - but I think it's a tremendously useful package and it's worth keeping it around (we've been using it in the past temporarily and I expect that we might use it again - IMHO better to keep it than doing install-uninstall dance every time that we actually need it).

File: packages/xstate-viz/package.json [hunk 3]
```diff
@@ -33,7 +33,6 @@
     "@types/styled-components": "^4.4.0",
     "react": "^16.12.0",
     "react-dom": "^16.12.0",
-    "tsdx": "^0.11.0",
```
Comment (Andarist) [inline packages/xstate-viz/package.json:36]: had to remove this as it has caused bizarre problems with `vue-jest` and we actually don't really even use `@xstate/viz` here, so better to remove this for now and try to bring it back when we need to

I've put my findings of the issue in Yarn's issue here: https://github.com/yarnpkg/yarn/issues/6285#issuecomment-751402589

**Review by Andarist — APPROVED** [review]

**Review by davidkpiano — APPROVED** [review]

**Pull request merged**